### PR TITLE
pkg/ciliumidentity: Add Cilium Identity Controller Operator hive cell

### PR DESCRIPTION
**Pull request by ovidiutirla**
Add the reconciler logic for reconciling CIDs, Pods, Namespaces.

The Cilium Identity API objects will be managed by the cilium-operator instead of the cilium-agent. The desired state for Cilium Identities is determined by the cilium-operator based on monitored events for Cilium Identities, Pods, Namespaces, and Cilium Endpoint Slices (if enabled). When pod updates occur for a unique label set derived from pod and namespace labels, the cilium-operator creates Cilium Identities.

Related CFP #27752
Draft full implementation: https://github.com/cilium/cilium/pull/33204

### CODE REVIEW

**Comment by ovidiutirla** [comment]
The reconciler and controller logic will be added in subsequent PRs.

/cc @dlapcevic

**Review by dlapcevic — APPROVED** [review]
lgtm

**Comment by ovidiutirla** [comment]
As per discussion with @pippolo84, I've added to this PR the reconciler logic along with a dummy controller that would be completed in a subsequent PR along with some metrics for operator managing CIDs.

File: operator/identitygc/crd_gc.go [hunk 1]
```diff
@@ -8,11 +8,12 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/cilium/cilium/operator/k8s"
```
Comment (pippolo84) [inline operator/identitygc/crd_gc.go]: Nit: we enforce the following import order

- Go stdlib packages
- external packages
- cilium/cilium packages

So this one should be in the last set (starting from line 17).
Actually, one of the linters should have complained about this, I'm not sure why it didn't.

File: operator/pkg/ciliumidentity/cache.go [hunk 2]
```diff
@@ -162,31 +168,20 @@ func (c *CIDUsageInPods) AssignCIDToPod(podName, cidName string) (string, int) {
...
-		return "", 0, errors.New("cilium identity not found in pods usage cache")
+		return "", 0, false
 	}
 	count := c.decrementUsage(cidName)
 	delete(c.podToCID, podName)
 
-	return cidName, count, nil
-}
-
-func (c *CIDUsageInPods) CIDUsedByPod(podName string) (string, bool) {
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache.go:180]: I think it would be better to break this commit into three separate commits:

- one to remove `CIDUsedByPod`
- one to add the logger to NewCIDState
- one to change `RemovePod` return type

File: operator/k8s/resources.go [hunk 3]
```diff
@@ -66,3 +66,16 @@ type Resources struct {
 	Pods                 resource.Resource[*slim_corev1.Pod]
 	Namespaces           resource.Resource[*slim_corev1.Namespace]
 }
+
+// HasCEWithIdentity returns true or false if the Cilium Endpoint store has
+// the given identity.
+// Once the CiliumEndpoint is migrated to use resources
+// the cepStore can be removed and used the local store.
```
Comment (pippolo84) [inline operator/k8s/resources.go]: I'm not sure I'm getting this part right: CiliumEndpoint in the operator is already using `Resource[T]` framework.

File: operator/pkg/ciliumidentity/cache_test.go [hunk 4]
```diff
@@ -5,15 +5,17 @@ package ciliumidentity
...
 	"reflect"
 	"sync"
 	"testing"
 
-	"github.com/cilium/cilium/pkg/labels"
-
 	"github.com/stretchr/testify/assert"
 
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging"
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cache_test.go]: Nit: move this imports with the the other ones below

File: operator/pkg/ciliumidentity/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/cilium/cilium/pkg/labels"
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Ditto

File: operator/pkg/ciliumidentity/cell.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/cell.go:1]: I don't think there's much value in breaking the controller logic into this commit (https://github.com/cilium/cilium/pull/33383/commits/82af79bdf79ce396ee5751676189ebb8f65716f7) and the next one (https://github.com/cilium/cilium/pull/33383/commits/7079ecb4fb6bca2fb683381f813c2b4030b8e6c9). Could you please merge them?

File: operator/pkg/ciliumidentity/reconciler.go [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+
+	minIDValue := idpool.ID(identity.GetMinimalAllocationIdentity(option.Config.ClusterID))
+	maxIDValue := idpool.ID(identity.GetMaximumAllocationIdentity(option.Config.ClusterID))
+	idAllocator := basicallocator.NewBasicIDAllocator(minIDValue, maxIDValue)
+
+	nsStore, _ := namespace.Store(ctx)
+	podStore, _ := pod.Store(ctx)
+	cidStore, _ := ciliumIdentity.Store(ctx)
+	cepStore, _ := ciliumEndpoint.Store(ctx)
+	cesStore, _ := ciliumEndpointSlice.Store(ctx)
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Is there a reason to swallow all the errors?

File: operator/pkg/ciliumidentity/controller.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (c *Controller) Start(hookContext cell.HookContext) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *Controller) Stop(hookContext cell.HookContext) error {
+	//TODO implement me
+	panic("implement me")
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/controller.go]: I know that this is just a placeholder, but there's no need to use a `panic`, a `return nil` is enough. Also, instead of the comment, I suggest to write something in the commit description to clarify that this controller will be linked later into the operator hive.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+			defer r.cidCreateLock.Unlock()
+			if e := r.handleCIDDeletion(prevCIDName); e != nil {
+				r.logger.Error("CID deletion failed when reconciling pod deletion",
+					logfields.CIDName, prevCIDName,
+					logfields.K8sPodName, podKey.String(),
+					logfields.Error, e)
+			}
+		}
+		return nil
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Since the current implementation of the GC is not the one proposed in PR https://github.com/cilium/cilium/pull/33380, this change should be thought to work with the existing implementation in order for this PR to be merged independently.

In case the pod is not found, we should simply early-returning to avoid allocating a CID.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+}
+
+func (r *reconciler) getNamespace(namespace string) (*slim_corev1.Namespace, error) {
+	nsLookupObj := &slim_corev1.Namespace{
+		ObjectMeta: slim_metav1.ObjectMeta{
+			Name: namespace,
+		},
+	}
+
+	ns, exists, err := r.nsStore.Get(nsLookupObj)
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Shouldn't we use `r.nsStore.GetByKey(resource.Key{Name: namespace})` instead?

File: operator/pkg/ciliumidentity/reconciler.go [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+		return fmt.Errorf("failed to allocate CID: %w", err)
+	}
+
+	r.desiredCIDState.Upsert(cidName, cidKey)
+
+	podName := podResourceKey(pod.Name, pod.Namespace).String()
+	prevCIDName, count := r.cidUsageInPods.AssignCIDToPod(podName, cidName)
+
+	r.logger.Info("CID allocated for pod",
+		logfields.K8sPodName, fmt.Sprintf("%s/%s", pod.Namespace, pod.Name),
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go:323]: I expect the CID allocation to happen pretty often, so it might be better to remove `fmt.Sprintf` in favor of a simple `+` concat (see https://github.com/cilium/cilium/issues/19571)

File: operator/pkg/ciliumidentity/reconciler.go [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+		logfields.CIDName, cidName,
+		logfields.OldIdentity, prevCIDName,
+		logfields.Labels, k8sLabels)
+
+	if len(prevCIDName) > 0 && count == 0 && !r.cidIsUsedInCEPOrCES(prevCIDName) {
+		_ = r.handleCIDDeletion(prevCIDName)
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Same here, right now the CID deletion is done through the "old" identity gc, so this is not needed.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+
+	if isNewCID {
+		r.queueOps.enqueueCIDReconciliation(cidResourceKey(cidName), 0)
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go:330]: Since `enqueueCIDReconciliation` is not implemented, let's add this once there will be actual logic to be run.

File: operator/pkg/ciliumidentity/controller.go [hunk 14]
```diff
@@ -66,7 +78,40 @@ func registerController(p params) {
...
+
+	return nil
+}
+
+func (c *Controller) enqueueCIDReconciliation(cidKey resource.Key, delay time.Duration) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *Controller) enqueuePodReconciliation(podKey resource.Key, delay time.Duration) {
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/controller.go]: I suggest to avoid adding those placeholders and simply add this logic in the next PRs where it is actually needed.

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+		//	initialStoreState: testCreateCIDObj(cidName, testLbsA),
+		//	expectedDelete:    cidName,
+		//},
+		//{
+		//	name:                "failed deletion",
+		//	cidKey:              cidResourceKey(cidName),
+		//	initialStoreState:   testCreateCIDObj(cidName, testLbsA),
+		//	errorDuringDeletion: true,
+		//	expectedQueueSize:   1,
+		//},
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go]: Same here, let's remove this for now, we'll think about re-adding these cases with the identity GC refactor PR.

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+		// TODO adapt once we decide how to handle CID deletion
+		//{
+		//	name:            "Pod not in store but in mapping",
+		//	newPod:          pod1,
+		//	initialPodToCID: map[resource.Key]string{podResourceKey(pod1.Name, ns1.Name): cidName},
+		//	expected: expectedState{
+		//		queueSize:     1,
+		//		existsInQueue: []string{cidName},
+		//	},
+		//},
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go]: Ditto

File: operator/pkg/ciliumidentity/reconciler.go [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+// the following:
+// 1. Nothing - If CID doesn't exist in both desired state cache and watcher's
+// store.
+// 2. Deletes CID - If CID only exists in the watcher's store, and it isn't used.
+// 3. Creates CID - If CID only exists in the desired state cache.
+// 4. Updates CID - If CIDs in the desired state cache and watcher's store are
+// not the same.
+func (r *reconciler) reconcileCID(cidResourceKey resource.Key) error {
+	cidName := cidResourceKey.Name
+	storeCIDObj, existsInStore, err := r.cidStore.CacheStore().GetByKey(cidResourceKey.Name)
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Why calling `CacheStore()` to get the underlying store?

File: operator/pkg/ciliumidentity/reconciler.go [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	var storeCID *cilium_api_v2.CiliumIdentity
+	if existsInStore {
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: I find this function pretty hard to read. We should try to break into two cases (like CID in store/CID not in store or alternatively CID in desired state/CID not in desired state) and call a separate helper to handle each one.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+		return nil
+	}
+
+	return r.updateCID(storeCID, cidKey)
+}
+
+func (r *reconciler) createCID(cidName string, cidKey *key.GlobalIdentity) error {
+	cidLabels := cidKey.GetAsMap()
+	selectedLabels, skippedLabels := identitybackend.SanitizeK8sLabels(cidLabels)
+	r.logger.Debug("Skipped non-kubernetes labels when labelling CID. All labels will still be used in identity determination", logfields.Labels, skippedLabels)
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go:207]: Shouldn't we guard this log checking the length of `skippedLables` first? Same for the call in `updateCID`.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+
+	cid.Labels = selectedLabels
+	cid.SecurityLabels = cidLabels
+
+	r.logger.Info("Updating CID", logfields.CIDName, cid.Name)
+
+	_, err := r.clientset.CiliumV2().CiliumIdentities().Update(context.TODO(), cid, metav1.UpdateOptions{})
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: We should use the reconciler context here, same for the other usages of the clientset.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the PR!

Did a first pass and left some comments. Let's address these before going ahead.

File: operator/identitygc/crd_gc.go (hunk 1 above)
Comment (ovidiutirla) [reply operator/identitygc/crd_gc.go]: This was formatted with `goimports` but adjusted now manually.

File: operator/pkg/ciliumidentity/cache.go (hunk 2 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache.go:180]: Done.

File: operator/pkg/ciliumidentity/cache_test.go (hunk 4 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cache_test.go]: done

File: operator/pkg/ciliumidentity/cell.go (hunk 6 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/cell.go:1]: done

File: operator/pkg/ciliumidentity/controller.go (hunk 14 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/controller.go]: These are required for the `newReconciler` as it expects to find the implementation of the controller. Removed the TODOs and panic though.

File: operator/pkg/ciliumidentity/controller.go (hunk 8 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/controller.go]: Done.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 5 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 7 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: checked for all the errors and returned in case of failure to initialize

File: operator/pkg/ciliumidentity/reconciler.go (hunk 17 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: updated to use cacheStore directly.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 18 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: No longer required if we are not using CacheStore()

File: operator/pkg/ciliumidentity/reconciler.go (hunk 20 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 11 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go:323]: I propose we maintain the current implementation. Utilizing `+` for string concatenation would necessitate creating additional strings due to Go's immutable strings. The linked issue suggests employing string builders or buffers, but incorporating them here might introduce unnecessary complexity and reduce code readability. We can always adjust log verbosity later if it becomes a performance concern.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 13 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go:330]: I think having the `enqueueCIDReconciliation` is an important part of this PR and should be kept to provide context on how we plan to make the operator allocate CIDs for pods. We can currently think about this as an abstract queue that will be later implemented.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 10 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: updated.

File: operator/pkg/ciliumidentity/reconciler_test.go (hunk 15 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go]: done

File: operator/pkg/ciliumidentity/reconciler_test.go (hunk 16 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 5 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go]: I still see `pkg/option` and `pkg/labels` imports at the same rows, maybe you forgot to push the changes?

File: operator/pkg/ciliumidentity/reconciler.go (hunk 11 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go:323]: > Utilizing + for string concatenation would necessitate creating additional strings due to Go's immutable strings.

That's correct, but very often the concatenation is still faster.

> We can always adjust log verbosity later if it becomes a performance concern.

Yep, I agree. No need to micro-optimize ahead of time, let's resolve this one :+1:

File: operator/pkg/ciliumidentity/reconciler.go (hunk 13 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go:330]: I took a look at the draft implementation in https://github.com/cilium/cilium/pull/33204
What about removing the interface and pass the reconciler a couple of callbacks, once for CID reconciliation and the other one for Pod reconciliation?
I am not seeing other implementations of the interface and the callbacks should allow equal flexibility (if not greater) in testing. WDYT?

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 21]
```diff
@@ -0,0 +1,721 @@
...
+				t.Errorf("Unexpected updateCID result: got %v, want %v", updateCID, tc.expectedUpdate)
+			}
+			if deleteCIDName != tc.expectedDelete {
+				t.Errorf("Unexpected deleteCIDName result: got %v, want %v", deleteCIDName, tc.expectedDelete)
+			}
+			if len(queueOps.fakeWorkQueue) != tc.expectedQueueSize {
+				t.Errorf("Unexpected fakeWorkQueue size: got %v, want %v", len(queueOps.fakeWorkQueue), tc.expectedQueueSize)
+			}
+
+			// TODO check for CID deletion too, adapt once we decide how to handle CID deletion
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go:428]: Could you please add a reference to the Identity GC PR where the discussion is taking place? That will give useful context to a future casual reader.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 18 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go]: I think that the function has not improved much removing the usage of the `CacheStore()`. Moreover, calling `handleCIDDeletion` is pointless, as in the current implementation, there is another cell taking care of the identities GC. Finally, the `handleCIDDeletion` placeholder should be removed and, in case, it will be added later in PR https://github.com/cilium/cilium/pull/33380 (and possible follow-ups).

**Review by pippolo84 — COMMENTED** [review]
Thanks for the follow up. :+1: 

Quickly reviewed the latest changes and left a couple of additional comments. I'll do another pass later.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 5 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: should be done now, not sure exactly how it got reformatted back to the old format.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 19 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go:207]: The current logger is configured for Debug only, but I believe we can utilize all labels for debugging purposes without requiring extra logic to verify their length.

File: operator/pkg/ciliumidentity/reconciler_test.go (hunk 21 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go:428]: Added a link to the other PR.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 13 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go:330]: IMO interfaces provide stronger contracts, type safety, and facilitate easier refactoring compared to callbacks. Even if we don't anticipate multiple reconciler implementations now, using an interface allows extensibility and maintainability for the future. Additionally, using callbacks could make the solution less readable.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 12 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 9 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 18 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: removed the `handleCIDDeletion`

File: operator/k8s/resources.go (hunk 3 above)
Comment (ovidiutirla) [reply operator/k8s/resources.go]: Indeed, you're right! Removed the comment.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 22]
```diff
@@ -0,0 +1,444 @@
...
+type reconciler struct {
+	logger          *slog.Logger
+	ctx             context.Context
+	clientset       k8sClient.Clientset
+	idAllocator     *basicallocator.BasicIDAllocator
+	desiredCIDState *CIDState
+	cidUsageInPods  *CIDUsageInPods
+	cidUsageInCES   *CIDUsageInCES
+	// Ensures no CID duplicates are created while allocating CIDs in parallel,
+	// and avoids race conditions when CIDs are being deleted.
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]:
```suggestion
	// Ensures no CID duplicates are created while allocating CIDs in parallel
	// (that is, when processing a pod event and a CID event concurrently)
	// and avoids race conditions when CIDs are being deleted.
```

File: operator/pkg/ciliumidentity/reconciler.go [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+	}
+
+	for _, ces := range r.cesStore.List() {
+		r.cidUsageInCES.ProcessCESUpsert(ces.Name, ces.Endpoints)
+	}
+}
+
+// syncPodsOnStartup ensures that all pods have a CID for their labels, and that
+// all non-used CIDs are deleted. Non used CIDs are those that aren't in use by
+// any of the pods and also don't exist in CESs (if CES is enabled).
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Shouldn't the part regarding deleting the CIDs be removed given that we'll stick to the old GC implementation, for now?

File: operator/pkg/ciliumidentity/reconciler.go [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+	cidName := cidResourceKey.Name
+	storeCID, existsInStore, err := r.cidStore.GetByKey(cidResourceKey)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		_ = r.makeIDAvailable(cidName)
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Here we should be able to differentiate between the cases:

- `makeIDAvailable` failed because the identity was not previously reserved in the allocator. Nothing to do in that case.
- `makeIDAvailable` failed for unexpected reasons. Here we should propagate the error

File: operator/pkg/ciliumidentity/reconciler.go [hunk 25]
```diff
@@ -0,0 +1,444 @@
...
+
+	return lastError
+}
+
+// reconcileCID ensures that the desired state for the CID is reached, by
+// comparing the CID in desired state cache and watcher's store and doing one of
+// the following:
+// 1. Nothing - If CID doesn't exist in both desired state cache and watcher's
+// store.
+// 2. Deletes CID - If CID only exists in the watcher's store, and it isn't used.
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: As above, this comment needs to be updated given that we are not deleting CIDs here.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 26]
```diff
@@ -0,0 +1,444 @@
...
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		_ = r.makeIDAvailable(cidName)
+		return nil
+	}
+
+	cidIsUsed := r.cidIsUsedInPods(cidName) || r.cidIsUsedInCEPOrCES(cidName)
+	if !existsInDesiredState {
+		return nil
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Couple of questions:

1) Why are we checking this here after calculating `cidIsUsed`? 

2) My understanding Is that the case `existsInDesiredState == false && cidIsUsed == true` is possible (or should be possible) only when the agent is managing CIDs together with the operator (cfr [here](https://docs.google.com/document/d/1Hcc_2mB9OOUxrqQgZ-gSYDPnLYE_If_TCzVbUGDOdGM/edit#heading=h.f1w8m5hi2mkd) in the CfP). Is my assumption correct? If so, shouldn't we mark the identity as "not available" in the basic ID allocator, to avoid the operator allocating that ID?

File: operator/pkg/ciliumidentity/reconciler.go [hunk 27]
```diff
@@ -0,0 +1,444 @@
...
+
+	if !cidIsUsed {
+		if existsInStore {
+			return nil
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go:174]: Might be worth adding a comment here stating that the identity GC will eventually delete the CID and that there's nothing more to do on the reconciler side, for now.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 28]
```diff
@@ -0,0 +1,444 @@
...
+		return nil, fmt.Errorf("namespace %q not found in store", namespace)
+	}
+
+	return ns, nil
+}
+
+func (r *reconciler) handleStoreCIDMatch(storeCIDs []*cilium_api_v2.CiliumIdentity) (string, error) {
+	if len(storeCIDs) == 0 {
+		return "", fmt.Errorf("store CIDs list is empty")
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: We already checked this before calling the function. Since `handleStoreCIDMatch` is called just once, there's no value in checking again `storeCIDs`.

File: operator/pkg/ciliumidentity/reconciler.go [hunk 29]
```diff
@@ -0,0 +1,444 @@
...
+			} else {
+				toDelete = false
+				selectedCIDName = cid.Name
+			}
+		}
+
+		if toDelete {
+			r.queueOps.enqueueCIDReconciliation(cidResourceKey(cid.Name), 0)
+		}
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: We are not deleting identities from the new controller, so is the call to `enqueueCIDReconciliation` (and the `toDelete` logic) necessary?

File: operator/pkg/ciliumidentity/reconciler.go (hunk 13 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go:330]: Let's leave it as is for now, we can still re-evaluate this later when the full implementation will be in.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the follow up.

Main concern is around some leftover logic that I think is not needed until we're gonna refactor the identity GC. Please be sure to rework every part that might assume a different GC implementation before requesting another review. Thanks! :pray:

File: operator/pkg/ciliumidentity/reconciler.go (hunk 23 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: updated the docs

File: operator/pkg/ciliumidentity/reconciler.go (hunk 25 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: updated the docs

File: operator/pkg/ciliumidentity/reconciler.go (hunk 24 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: I've added a log message to track this, I don't believe we need to propagate the error from reconcileCID.
If it would throw an error, we'll simply re-queue the cid reconciliation and it'll fail again. 

The potential failure reasons:
1) cidName is not an int (shouldn't happen)
2) cid is not in the range (shouldn't happen)
3) cid is there already so we shouldn't anyway propagate it (shouldn't happen)

File: operator/pkg/ciliumidentity/reconciler.go (hunk 26 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: 1. refactored the if statements to improve redability
2. Added your suggestion. This scenario is unlikely to occur in real situations because the order of events should prevent it. The operator typically reacts to the creation of a pod object before the agent has a chance to process the corresponding pod event.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 28 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: done

File: operator/pkg/ciliumidentity/reconciler.go (hunk 29 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: refactored the method as we no longer need to reconcile all CIDs if operator managing CIDs is no longer deleting the CIDs.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 27 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go:174]: added to the methods docs.

File: operator/pkg/ciliumidentity/reconciler.go (hunk 24 above)
Comment (pippolo84) [reply operator/pkg/ciliumidentity/reconciler.go]: Let's go with the warning, then! :+1:

File: operator/pkg/ciliumidentity/reconciler.go [hunk 30]
```diff
@@ -0,0 +1,438 @@
...
+				return err
+			}
+			err = r.idAllocator.Allocate(idpool.ID(id))
+			if err != nil {
+				return err
+			}
+
+		}
+		return nil
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler.go]: Nonblocking: using an early return here makes the code a bit nicer

```suggestion
	if !existsInDesiredState {
		if !cidIsUsed {
			return nil
		}

		r.cidCreateLock.Lock()
		defer r.cidCreateLock.Unlock()

		id, err := r.idAllocator.ValidateIDString(cidName)
		if err != nil {
			return err
		}
		return r.idAllocator.Allocate(idpool.ID(id))
	}
```

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 31]
```diff
@@ -0,0 +1,715 @@
...
+					actualCount := reconciler.cidUsageInCES.cidUsageCount[cid]
+					if actualCount != expectedCount {
+						t.Errorf("Expected CID %d usage count to be %d, but got %d", cid, expectedCount, actualCount)
+					}
+				}
+			} else if len(reconciler.cidUsageInCES.cidUsageCount) != 0 {
+				t.Errorf("Expected no CIDs to be used when CES is disabled, but found %d", len(reconciler.cidUsageInCES.cidUsageCount))
+			}
+		})
+	}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go]: Since we are using table driven sub-tests and we just need to compare two maps, we can avoid differentiating the test logic with an if an do this instead:

```go
t.Run(tc.name, func(t *testing.T) {
	...

	reconciler.syncCESsOnStartup()

	if !reflect.DeepEqual(reconciler.cidUsageInCES.cidUsageCount, tc.expectedCIDs) {
		t.Errorf("Expected CID usage count map to be %v, but got %v", tc.expectedCIDs, reconciler.cidUsageInCES.cidUsageCount)
	}
})
```

(for this to work you also have to change the `expectedCIDs` in the first subtest to be `map[int64]int{}`, otherwise you can use something like https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#EquateEmpty)

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 32]
```diff
@@ -0,0 +1,715 @@
...
+}
+
+func TestHandleStoreCIDMatch(t *testing.T) {
+	testCases := []struct {
+		name          string
+		cidList       []*capi_v2.CiliumIdentity
+		expectedCID   string
+		expectedError bool
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go]: There are no sub-tests where we expect an error, so we can remove this and simply check for no error at every run.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
This looks better, thanks! :100: 

Just a final couple of comments on the tests (nothing major) and a non-blocking suggestion to make the reconciling main function a bit nicer.

File: operator/pkg/ciliumidentity/reconciler_test.go (hunk 31 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go]: done

File: operator/pkg/ciliumidentity/reconciler_test.go (hunk 32 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go]: removed expectedError

File: operator/pkg/ciliumidentity/reconciler.go (hunk 30 above)
Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler.go]: refactored

File: operator/pkg/ciliumidentity/reconciler_test.go [hunk 33]
```diff
@@ -0,0 +1,716 @@
...
+			//if tc.cesEnabled {
+			//	for cid, expectedCount := range tc.expectedCIDs {
+			//		actualCount := reconciler.cidUsageInCES.cidUsageCount[cid]
+			//		if actualCount != expectedCount {
+			//			t.Errorf("Expected CID %d usage count to be %d, but got %d", cid, expectedCount, actualCount)
+			//		}
+			//	}
+			//} else if len(reconciler.cidUsageInCES.cidUsageCount) != 0 {
+			//	t.Errorf("Expected no CIDs to be used when CES is disabled, but found %d", len(reconciler.cidUsageInCES.cidUsageCount))
+			//}
```
Comment (pippolo84) [inline operator/pkg/ciliumidentity/reconciler_test.go]: Leftover commented code

Comment (ovidiutirla) [reply operator/pkg/ciliumidentity/reconciler_test.go]: removed

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks! :rocket:

**Review by pchaigno — APPROVED** [review]
Approving for the 2 lines of "CLI" change (logfield).
